File: examples/integration/julia/bin/parse_words.jl
using LinkedSpecJulia
using JSON3

function main(arguments)
    try
        length(arguments) >= 2 || throw(ArgumentError(
            "Usage: parse_words.jl GRAMMAR INPUT [INPUT ...]",
        ))
        # The managed Julia wrapper changes cwd; use an explicit application root.
        app_root = normpath(joinpath(@__DIR__, ".."))
        loaded = load_and_compile_spec(
            path_spec_request(arguments[1]),
            SpecLoadOptions(; cwd = app_root),
        )
        engine = create_engine(loaded)
        for input in arguments[2:end]
            # Compile once, then parse independent inputs in this Julia process.
            result = runtime_execute(engine, input; top_rule = "Top")
            println(JSON3.write(result.value))
        end
        return 0
    catch error
        if error isa InterruptException || error isa OutOfMemoryError || error isa StackOverflowError
            rethrow()
        end
        record = if error isa SpecPipelineException
            to_json(error)
        elseif error isa RuntimeInterpreterException
            merge(Dict("type" => "runtime_error"), to_json(error))
        else
            Dict("type" => "consumer_error", "detail" => sprint(showerror, error))
        end
        println(stderr, JSON3.write(record))
        return 1
    end
end

exit(main(ARGS))

File: examples/integration/lua/bin/parse_words.lua
-- The managed wrapper supplies the checkout root; select this ABI's retained products.
local source_root = assert(os.getenv("LINKEDSPEC_REPO_ROOT"), "LINKEDSPEC_REPO_ROOT is required")
local native_root = assert(os.getenv("LINKEDSPEC_LUA_NATIVE_ROOT"), "LINKEDSPEC_LUA_NATIVE_ROOT is required")
package.path = source_root .. "/lua/src/?.lua;" .. source_root .. "/lua/src/?/init.lua"
package.cpath = native_root .. "/?.so"

local linkedspec = require("linkedspec")
local filesystem = require("linkedspec_filesystem_native")
local json = linkedspec.json

local function main()
  if #arg < 2 then
    error("Usage: parse_words.lua GRAMMAR INPUT [INPUT ...]", 0)
  end
  local loaded = linkedspec.load_and_compile_spec(
    linkedspec.path_spec_request(arg[1]),
    linkedspec.spec_load_options({ cwd = filesystem.current_directory(), search_roots = {} })
  )
  local engine = loaded:create_engine()
  for index = 2, #arg do
    -- Compile once, then execute independent inputs in this Lua process.
    local result = linkedspec.runtime_parse(engine, arg[index], { top_rule = "Top" })
    local value = result.value
    if value == nil then value = json.null end
    io.stdout:write(json.encode(value), "\n")
  end
end

local ok, failure = pcall(main)
if not ok then
  local record
  if linkedspec.is_spec_pipeline_error(failure) then
    record = linkedspec.spec_pipeline_error_to_json(failure)
  elseif linkedspec.is_runtime_interpreter_error(failure) then
    record = linkedspec.interpreter.to_json(failure)
    record.type = "runtime_error"
  else
    record = json.harray({ type = "consumer_error", detail = tostring(failure) })
  end
  io.stderr:write(json.encode(record), "\n")
  os.exit(1)
end
